main.cpp

```diff
@@ -90,315 +89,166 @@ struct Point
 public:    
-    FloatType(float value_) : value( new float(value_) ) {}
-    ~FloatType()
-    { 
-        delete value; 
-        value = nullptr;
-    }
+
+    using CurrentType = TemplatedType;
+    
+    Numeric(CurrentType value_) : value( new CurrentType(value_) ) {}
     // Operator overloading
-    FloatType& operator+=(float rhs);
-    FloatType& operator-=(float rhs);
-    FloatType& operator*=(float rhs);
-    FloatType& operator/=(float rhs);
-    // 4 power functions with different exponent type
-    FloatType& powInternal(const float exp);
-    FloatType& pow(const float exp); 
-    FloatType& pow(const IntType& itExp);
-    FloatType& pow(const FloatType& ftExp);
-    FloatType& pow(const DoubleType& dtExp);    
-    // convert UDT to primitive type
-    operator float() const
+    Numeric& operator+=(CurrentType rhs)
     {
-        return *value;
+        *value += rhs;
+        return *this;
     }
-    // 
-    FloatType& apply(std::function<FloatType&(float&)> myFunc);
-    FloatType& apply(void(*myFunc)(float&));
-}; 
-// Operator overloading
-FloatType& FloatType::operator+=(float rhs) 
-{
-    *value += rhs;
-    return *this;
-}
-FloatType& FloatType::operator-=(float rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-FloatType& FloatType::operator*=(float rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-FloatType& FloatType::operator/=(float rhs)
-{
-    *value /= rhs;
-    return *this;
-}
-// Apply functions
-FloatType& FloatType::apply(std::function<FloatType&(float&)> myFunc)
-{
-    if( myFunc )
-        return myFunc(*value);
-    return *this;
-}
-FloatType& FloatType::apply(void(*myFunc)(float&))
-{
-    if( myFunc )
-        myFunc(*value);
-    return *this;
-}
-// =============================================================
-//                  DoubleType object definition
-// =============================================================
-struct DoubleType
-{
-private:
-    double* value;
-public:   
-    DoubleType(double value_) : value( new double(value_) ) {}
-    ~DoubleType()
-    { 
-        delete value; 
-        value = nullptr;
+    Numeric& operator-=(CurrentType rhs)
+    {
+        *value -= rhs;
+        return *this;
     }
-    // Operator overloading
-    DoubleType& operator+=(double rhs);
-    DoubleType& operator-=(double rhs);
-    DoubleType& operator*=(double rhs);
-    DoubleType& operator/=(double rhs);
+    Numeric& operator*=(CurrentType rhs)
+    {
+        *value -= rhs;
+        return *this;
+    }
+    Numeric& operator/=(CurrentType rhs)
+    {
+        if (std::is_same<CurrentType, int>::value) 
+        {
+            // Only way to silence all the warnings for comaparison (-Wfloat-equal)
+            if( std::abs(rhs) <= std::numeric_limits<CurrentType>::epsilon() )
+                *value /= rhs;
+            else    
+                std::cout << "Division by zero not allowed with integers!!" << std::endl;
+            return *this;       
+        }
```

Comment:
I could be wrong, but I think you just need this:
```
if( std::is_same<CurrentType, int>::value && rhs == 0 )
{
    std::cout << "Division by zero not allowed with integers!!" << std::endl;
    return *this;     
}
else if( std::abs(rhs) <= std::numeric_limits<CurrentType>::epsilon() )
{
    std::cout << "Warning: Floating point division by zero!"
}
*value /= rhs;
return *this;
```
- [ ] investigate this revision

Reply:
The condition `rhs == 0` would give the same warning I wanted to silence using `epsilon`. Anyway, I found a way around.